File: deepest-leaves-sum/deepest-leaves-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        self.max_depth = 0
        
        def dfs(node=root, d=0):
            if not node: return
            
            if self.max_depth < d:
                self.max_depth = d
                self.res = 0
            if d == self.max_depth: 
                self.res += node.val
            
            dfs(node.left, d+1)
            dfs(node.right, d+1)
            
        dfs()
        return self.res

File: deepest-leaves-sum/deepest-leaves-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        int res = 0, max_depth = 0;
        
        function<void(TreeNode*, int)> dfs = [&] (TreeNode* node, int depth){
            if (!node) return;
            if (max_depth < depth) {
                max_depth = depth;
                res = 0;
            };
            if (depth == max_depth){
                res += node->val;
            }
            dfs(node->left, depth + 1);
            dfs(node->right, depth +1);
        };
            
        dfs(root, 0);
        return res;
            
    }

};